codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java

```diff
@@ -195,14 +195,26 @@ private FlowState connectionState() {
     }
 
     private static FlowState state(Http2Stream stream) {
-        checkNotNull(stream, "stream");
-        return stream.getProperty(FlowState.class);
+        if (stream instanceof FastPropertyAccess) {
```

Comment:
Hmm it would be nice to not have to constantly do this when we know it will evaluate to true. `instanceof` may be fast but it still has to traverse the type hierarchy....and it isn't free.

I wonder if it would be cleaner to do the following:
1. Create a read only public interface for `LocalFlowState` and `RemoteFlowState`.
2. Add a `LocalFlowState localFlowState();` and `RemoteFlowState remoteFlowState();` to `Http2Stream`.
3. Have `Http2LocalFlowController` and `Http2RemoteFlowController` expose factory methods which can generate new `LocalFlowState` and `RemoteFlowState` objects.
4. The `Http2Connection` interface documents that it is responsible for using these factories when a stream becomes "added" and before listeners are notified.

This way we shouldn't have to do any instance of checks. The `DefaultHttp2LocalFlowController` and `DefaultHttp2RemoteFlowController` should be able to cast to whatever private class they define in their factories.  I understand we are using the more generic `Connection.Listener` methods today...but flow control is a part of the HTTP/2 spec and it must be supported...it may be worth explicitly having provisions for it rather than hooking into these generic methods.


Another potential improvement in this area is the `DefaultHttp2LocalFlowController` and `DefaultHttp2RemoteFlowController ` implementations of the `*FlowState` interface could also be more conservative when the "full" set of state is needed.  For example a stream should not need the full amount of flow control state until it is OPENed, and then can use a reduced set of state when the stream is CLOSED.  So for example if a stream is in IDLE and in the priority tree (i.e. only PRIORITY frames have been exchanged), or the stream has been closed...but it has dependents in the priority tree which are not closed then the stream will remain in the priority tree.